### PR TITLE
Restructure ZnTrack to use Class Inhertiance instead of `@Node` decorator

### PR DESCRIPTION
**Pull request by PythonFZ**
This PR will break backwards compatibility. A Guide to convert existing Nodes will be provided.

# TODOs
- [x] Test Code
- [x] Move Examples from docs to example directory
- [x] Add binder to the docs so people can interactively try out ZnTrack
- [x] Check for missing functionality
- [x] Fix class dependencies
- [x] Update Docs
- [x] Write migrations Guide + fix warning in the `@Node`
- [ ] Gather Feedback
- [ ] Add Warnings when loading / saving options that should not be changed
- [x] Fix named Nodes
- [x] Fix default values
- [x] fix list dependencies
- [x] fix JupyterNotebook Mode #162 
- [x] `dvc.params` should be `zn.params` and `dvc.params` should also allow for a file parameter to be constistent
- [ ] split `dvc.deps` for paths and `zn.deps` for Nodes ?
- [ ] ... missing points to be added

# Summary
The new API is more inline with Python standards and will be based on class inheritance.
- `Node().run` has been replaced by `Node().run_and_save` whilst the user will still overwrite the `Node.run` method. This allows for easier testing and highlights that the results are saved afterwards.
- `Node(load=True)` is replaced by the classmethod `Noad.load(name=<...>)`
- `Node.__call__` is replaced by `Node.write_dvc()` This gives the user more flexibility and highlights that a DVC file is being written. On the oder hand the `__init__` currently must have default values, because otherwise the loading will not work as expected
- `@Node` is replaced by `class MyNode(zntrack.Node)`
- `dvc.result` was removed in favour of `zn.outs()`

# Example

```py
class HelloWorld(zntrack.Node):
    param1 = zn.params()
    param2 = zn.params()
    result = zn.outs()
    
    def __init__(self, param1=None, param2=None):
        self.param1 = param1
        self.param2 = param2
    
    def run(self):
        self.result = self.param1 + self.param2

# Write the DVC file and save the parameters
hello_world = HelloWorld("Lorem Ipsum", 42)
hello_world.write_dvc()

# Load the Node, run it and save it (usually executed by DVC)
HelloWorld.load().run_and_save()
```

### CODE REVIEW

File: zntrack/descriptor/descriptor.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+class Metadata:
+    dvc_option: str
+    zntrack_type: str
+
+    @property
+    def dvc_args(self):
+        return self.dvc_option.replace("_", "-")
+
+
+class Descriptor:
```
Comment (SamTov) [inline zntrack/descriptor/descriptor.py:17]: I find it hard to know exactly what a descriptor is as it seems specific to some context. Having a description here or in the header doc string (which is missing) would be very helpful.

**Review by SamTov — APPROVED** [review]
I just had one comment in regard to being able to follow and understand each module and class. I know it may sound repeated but I think in one section there were a lot of classes without doc strings and particularly for a package that is quite unique and requires some background, this makes it very hard to understand the purpose of each class or method. 

To the extent that I read the code I though it looked great and couldn't see anything I would change or do differently. In terms of understanding, i read the descriptions where they existed and tried to follow function by function but I am not familiar enough to make complete comments on how functional it all is.

File: zntrack/descriptor/descriptor.py (hunk 1 above)
Comment (PythonFZ) [reply zntrack/descriptor/descriptor.py:17]: It is a Python standard term https://docs.python.org/3/howto/descriptor.html

Comment (PythonFZ) [reply zntrack/descriptor/descriptor.py:27]: I added some in 8bf456b

**Pull request merged**